[16.0][IMP] base_geoengine: Include edit button in Popup

anusriNPS opened the pull request:
Added edit button in popup of record shown in geoengine view which allows user to edit records by opening new tab in browser window.

anusriNPS committed:
[16.0][IMP] base_geoengine: Include edit button in Popup

   Added edit button in popup of record shown in geoengine
view which allows user to edit values of record from
geoengine view.

cparadis-impressfoods reviewed (approved):
LGTM!

quirino95 commented on base_geoengine/static/src/js/views/geoengine/geoengine_renderer/geoengine_renderer.xml:
```diff
@@ -14,6 +14,12 @@
         </div>
         <div id="popup" class="ol-popup">
+             <div
+                t-if="isGeoengineAdmin"
+                id="popup-editor"
+                class="ol-popup-editor text-primary"
+                t-on-click="onEditButtonClicked"
+            />
```
I have some observations about this:
- Is really GeoenginAdmin check needed? I think is more correct that the check should be done on the work order itself (user could be GeoengineAdmin but not allowed to edit a specific WO, how is this situation managed?);
- When moving the cursor over the button, it doesn't give me the impression that it's clickable. Please could you improve this by adding a `clickable` attribute or changing the `div` into a `button` or something similar?
- When clicking on `edit` button, the form view just opened is "full screen", hiding the upper navigation bar, and when going back, map view is reset;
- **suggestion**: clicking the "open" button, i see the same form that I have by clicking on "edit" button. What if the open button opens form view in edit mode (depending of access rules of course)? In that case, "edit" button would become obsolete, but the interface should become more clear;

HekkiMelody replied:
Re: the first point, this is a generic geoengine module, it could be used with any object, not just fieldservice. But in general, I agree with checking permissions on the object itself if possible and not requiring admin rights on the Geoengine app.

If it's too complicated, Odoo probably already checks permissions in other ways so I'd rather leave both buttons for everyone, at least as a first implementation.

anusriNPS replied:
1. Regarding edit button, icon is shown as clickable attribute when hovered around it. Edit button opens url in new tab with navigation bar and also restores history in geoengine view.
2. Regarding permission on object level:
     a. Tried to hide the editbutton using ormservice and reading permission for record from backend during onwillstart() in geoengineRenderer. However, each record related information is attached in geoengineRenderer only when displayPopupRecord() method is called and geoenginerecord is mounted based on information passed from recordPanel template. Component is getting destroyed while trying to fetch specific record from geoengineRenderer setup. So, could not hide the button completely. This would be best option for this case but could not achieved.
     b. Currently, if a model has a field named "can_write", if passed in info_box template from geoengine view definition, record is validated based on can_write field and allows user to edit the record. If not, user is notified that sufficient permissions are not available to edit the record. If info_box template does not have "can_write" field, all users is allowed to edit the record.(default behaviour).
     
Note:  Without approach "b", by default even when user is allowed to edit the record, permissions are validated in backend and user is blocked in case the user does not have sufficient permissions.

IMO, I would like to stick not using approach "b" as already backend permissions are validated even without exposing new field in model for this purpose which seems to be overhead. 

Please let me know your views.

quirino95 replied:
Changes on the edit button now are ok.
About permissions, since now the form opens in a new tab, is ok for me to remove new checks added in this PR and stay with the "default" ones when opening the new tab.

anusriNPS replied:
Removed checking permission based on "can_write" field introduced from model and preserved default behaviour for permission check of geoengine views for edit button.

HekkiMelody commented on base_geoengine/static/src/js/views/geoengine/geoengine_controller/geoengine_controller.esm.js:
```diff
@@ -64,6 +64,31 @@ export class GeoengineController extends Component {
+     * @param {*} resId
+     */
+    async editRecord(resModel, resId) {
+        var web_base_url = window.origin;
+        var web_url = window.location.href;
+        var searchParams = new URLSearchParams(web_url.split("#")[1]);
+        var action_url =
+            web_base_url +
+            "/web#id=" +
+            String(resId) +
+            "&view_type=form&model=" +
+            resModel +
+            "&menu_id=" +
+            searchParams.get("menu_id");
+        this.actionService.doAction({
+            type: "ir.actions.act_url",
+            target: "_blank",
+            url: action_url,
+        });
```
**chore**: not at all a fan of building the URL manually, but if there's no other easy way to do it, that's fine.

I would suggest building the `action_url` in a less "manual" way, like in this module https://github.com/OCA/web/blob/16.0/web_widget_open_tab/static/src/js/open_tab_widget.esm.js#L12-L25

Also, please use `const` and `let` instead of `var`.

anusriNPS replied:
Updated as suggested

HekkiMelody reviewed (commented):
Small code cleanup suggestion, otherwise seems ok.

HekkiMelody reviewed (approved):
Code review, LGTM

HekkiMelody commented:
@OCA/geospatial-maintainers We've been using this successfully for a few months now. I would be grateful if you could review this. Thanks

legalsylvain commented:
No review, but as it is validated by 3 people. merging.

/ocabot merge minor

thanks for your contribution.

HekkiMelody commented:
I see that the error is due to

```
2026-01-26 11:26:30,614 334 INFO odoo odoo.addons.base_geolocalize.models.base_geocoder: openstreetmap nominatim service called 
2026-01-26 11:26:30,615 334 WARNING odoo odoo.addons.base_geolocalize.models.base_geocoder: Request to openstreetmap failed.
Code: 403
Content: b'<html>\n<head>\n<title>Access blocked</title>\n</head>\n<body>\n<h1>Access blocked</h1>\n\n<p>You have been blocked because you have violated the\n<a href="https://operations.osmfoundation.org/policies/nominatim/">usage policy</a>\nof OSM\'s Nominatim geocoding service. Please be aware that OSM\'s resources are\nlimited and shared between many users. The usage policy is there to ensure that\nthe service remains usable for everybody.</p>\n\n<p>Please review the terms and make sure that your\nsoftware adheres to the terms. You should in particular verify that you have set a\n<b>custom HTTP referrer or HTTP user agent</b> that identifies your application, and\nthat you are not overusing the service with massive bulk requests.\nIf you do that then the block usually goes away after some hours of inactivity.</p>\n\n<p>Please be aware that we are providing this service as-is and without any\nguarantees. We cannot provide support. If you need that, please have a look\nat comercial alternatives (also listed on the usage policy page) or set up\nyour own instance of Nominatim.</p>\n</body>\n</head>\n' 
2026-01-26 11:26:30,617 334 INFO odoo odoo.addons.geoengine_base_geolocalize.tests.test_geoengine_partner: ====================================================================== 
2026-01-26 11:26:30,617 334 ERROR odoo odoo.addons.geoengine_base_geolocalize.tests.test_geoengine_partner: ERROR: TestGeoenginePartner.test_geo_localize
Traceback (most recent call last):
  File "/opt/odoo/addons/base_geolocalize/models/base_geocoder.py", line 95, in _call_openstreetmap
    result = response.json()
  File "/opt/odoo-venv/lib/python3.10/site-packages/requests/models.py", line 900, in json
    return complexjson.loads(self.text, **kwargs)
  File "/opt/odoo-venv/lib/python3.10/site-packages/simplejson/__init__.py", line 514, in loads
    return _default_decoder.decode(s)
  File "/opt/odoo-venv/lib/python3.10/site-packages/simplejson/decoder.py", line 386, in decode
    obj, end = self.raw_decode(s)
  File "/opt/odoo-venv/lib/python3.10/site-packages/simplejson/decoder.py", line 416, in raw_decode
    return self.scan_once(s, idx=_w(s, idx).end())
simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

```

we may be hitting openstreetmaps a bit too often? We may have to refactor the tests to not call OSM, or skip some tests in case of `Access Denied`

anusriNPS commented:
> I see that the error is due to
> 
> ```
> 2026-01-26 11:26:30,614 334 INFO odoo odoo.addons.base_geolocalize.models.base_geocoder: openstreetmap nominatim service called 
> 2026-01-26 11:26:30,615 334 WARNING odoo odoo.addons.base_geolocalize.models.base_geocoder: Request to openstreetmap failed.
> Code: 403
> Content: b'<html>\n<head>\n<title>Access blocked</title>\n</head>\n<body>\n<h1>Access blocked</h1>\n\n<p>You have been blocked because you have violated the\n<a href="https://operations.osmfoundation.org/policies/nominatim/">usage policy</a>\nof OSM\'s Nominatim geocoding service. Please be aware that OSM\'s resources are\nlimited and shared between many users. The usage policy is there to ensure that\nthe service remains usable for everybody.</p>\n\n<p>Please review the terms and make sure that your\nsoftware adheres to the terms. You should in particular verify that you have set a\n<b>custom HTTP referrer or HTTP user agent</b> that identifies your application, and\nthat you are not overusing the service with massive bulk requests.\nIf you do that then the block usually goes away after some hours of inactivity.</p>\n\n<p>Please be aware that we are providing this service as-is and without any\nguarantees. We cannot provide support. If you need that, please have a look\nat comercial alternatives (also listed on the usage policy page) or set up\nyour own instance of Nominatim.</p>\n</body>\n</head>\n' 
> 2026-01-26 11:26:30,617 334 INFO odoo odoo.addons.geoengine_base_geolocalize.tests.test_geoengine_partner: ====================================================================== 
> 2026-01-26 11:26:30,617 334 ERROR odoo odoo.addons.geoengine_base_geolocalize.tests.test_geoengine_partner: ERROR: TestGeoenginePartner.test_geo_localize
> Traceback (most recent call last):
>   File "/opt/odoo/addons/base_geolocalize/models/base_geocoder.py", line 95, in _call_openstreetmap
>     result = response.json()
>   File "/opt/odoo-venv/lib/python3.10/site-packages/requests/models.py", line 900, in json
>     return complexjson.loads(self.text, **kwargs)
>   File "/opt/odoo-venv/lib/python3.10/site-packages/simplejson/__init__.py", line 514, in loads
>     return _default_decoder.decode(s)
>   File "/opt/odoo-venv/lib/python3.10/site-packages/simplejson/decoder.py", line 386, in decode
>     obj, end = self.raw_decode(s)
>   File "/opt/odoo-venv/lib/python3.10/site-packages/simplejson/decoder.py", line 416, in raw_decode
>     return self.scan_once(s, idx=_w(s, idx).end())
> simplejson.errors.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
> ```
> 
> we may be hitting openstreetmaps a bit too often? We may have to refactor the tests to not call OSM, or skip some tests in case of `Access Denied`

Will check on the issue

anusriNPS committed:
[16.0][IMP] geoengine_base_geolocalize: Using mock data for OpenStreetMap

   Avoiding direct call to openstreetmap which results in
Access denied (403) due to resource usage limits and hence mocking data
for openstreetmap url as part of geoengine partner testcases.

quirino95 reviewed (approved):
LGTM

Thank you for the fix!

HekkiMelody reviewed (approved):
Code review, LGTM for the test mock too.

Thank you!

HekkiMelody commented:
@OCA/geospatial-maintainers we've fixed the issue with the flaky tests (even though they were out of scope for this PR) could you try the merge again? Thank you

We can forward port both changes to the newest versions too

legalsylvain commented:
thanks for the contribution ! 

/ocabot merge minor

HekkiMelody commented:
Thank you @legalsylvain !
